lib/internal/http2/core.js

```diff
@@ -295,25 +296,14 @@ function tryClose(fd) {
   fs.close(fd, (err) => assert.ifError(err));
 }
 
-// Called to determine if there are trailers to be sent at the end of a
-// Stream. The 'getTrailers' callback is invoked and passed a holder object.
-// The trailers to return are set on that object by the handler. Once the
-// event handler returns, those are sent off for processing. Note that this
-// is a necessarily synchronous operation. We need to know immediately if
-// there are trailing headers to send.
+// Called when the Http2Stream has finished sending data and is ready for
+// trailers to be sent. This will only be called if the { hasOptions: true }
+// option is set.
 function onStreamTrailers() {
   const stream = this[kOwner];
   if (stream.destroyed)
-    return [];
-  const trailers = Object.create(null);
-  stream[kState].getTrailers.call(stream, trailers);
-  const headersList = mapToHeaders(trailers, assertValidPseudoHeaderTrailer);
-  if (!Array.isArray(headersList)) {
-    stream.destroy(headersList);
-    return [];
-  }
-  stream[kSentTrailers] = trailers;
-  return headersList;
+    return;
+  process.nextTick(emit, stream, 'trailers-ready');
```

Comment:
What happens if the event listener never calls `.trailers()`?

Reply:
Then it doesn't close and things hang until it does. Exactly the same as if `end()` is never called on an HTTP `response` object.

The stream can be closed using either `trailers()` or `close()` on the `Http2Stream`.

That should definitely be emphasized more in the docs.

Reply:
Okay … I don’t see any way to do that better, I guess, but could we at least make it an early error if there are no listeners for the event at this point?

Reply:
potentially, yes, but doing so would lock us in to having those trailers available right when that event is triggered when it could be any time after that event. Will be stewing on this more.

(this, btw, is the part I've been stewing on for a while with this which is why it took so long to get this done. I haven't been able to come up with a better solution.)

Reply:
I meant basically adding `if (stream.listenerCount('trailers-ready') === 0) throw new Error();` here, right before this line – nothing fancier than that

Reply:
Or just simply `close()` rather than throw. That's certainly possible. The `trailers-ready` (or whatever we use) is a marker that says trailers can be sent at any time after this. If the user is not listening for that, then they cannot reasonably know when to proceed with sending the trailers anyway.